### ISSUE TITLE
These functions should have identical hashes

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/unison#3015]
These two versions of `isRight` look identical to me but somehow have different hashes. How?

<img width="303" alt="image" src="https://user-images.githubusercontent.com/11074/161073898-82520a59-daed-44e4-9ea4-3a12e889483a.png">

Here's where you can get the codebase with just the two definitions in it:

```
pull git@github.com:pchiusano/sketches:#95rm7houk8
```

**Comment by pchiusano** [comment]
If I had to take a guess, I'd say it's that there's actually two type signatures on the function. That is, the AST is actually (`(e : sig1) : sig2`) but the display normalizes for this while hashing does not. It could also be something with hidden ability variables, though this function should be pure and there's nothing higher order.

@mitchellwrosen could you confirm that the logic is the following - after a top level definition gets its type inferred, that inferred type should *replace* any top-level `Ann` on the term (or add a new top-level `Ann` if none exists). It should not just be added as *another* `Ann` node on top of one that already exists. If the user provides a type signature, the inferred type will match what they provided (possibly with elaborated ability variables), so the inferred type can and should always be used.

I played with this some more and found ANOTHER variant with somehow a different hash - this one I left off the signature and just let it infer.

<img width="435" alt="image" src="https://user-images.githubusercontent.com/11074/161076584-2b95772a-d72c-442e-bc97-c7dba629f06a.png">

```
pull git@github.com:pchiusano/sketches:#e5c7cr44li
```

If my hypothesis is correct, I think that this issue could also be behind some super annoying weirdness I've seen with update propagation. Like I make a type-preserving or type isomorphic change which should propagate fully automatically, but it doesn't.

If there's a nested type signature in there sometimes, that could explain it - during propagation, the type signature is rewritten based on the elaborated patch, but that's only the outermost type signature. If there's a (redundant) inner type signature just below it, that won't get rewritten, and the definition won't typecheck because the outer signature `sig2` in (`(e : sig1) : sig2`) differs from the inner signature.

**Comment by pchiusano** [comment]
@mitchellwrosen I assigned to you. I don't think it's urgent, though.

**Comment by mitchellwrosen** [comment]
Thanks, I can look into this more when I find some time, or if/when it becomes urgent.

I did play around a little bit at the command line and couldn't come up with a way of getting a second type ascription like `(e : sig1) : sig2)` as you suspected, but I did manage to figure out the full types for your `isRight1` / `isRight2` / `isRight2`:

```unison
isRight1 : forall a b e. base.Either a b ->{e} Boolean

isRight2 : forall e a b. base.Either a b ->{e} Boolean

isRight3 : forall b a. base.Either a b ->{} Boolean
```

I'd be curious to hear if `isRight1` and `isRight2` are supposed to be equivalent, because type variable order is supposed to be normalized (these would be different types in Haskell because we have `-XTypeApplications`).

About `isRight3`, I think that might be a regression in inference. Here's another similar example.

```unison
id1 x = x

id2 : a ->{} a
id2 x = x
```

These two have the same hash, but I had thought `id1` is meant to infer as ability-polymorphic.

**Comment by pchiusano** [comment]
Ah. I think that explains it, then. We should normalize for quantifier order.

isRight3 is the correct type to infer, I’d say. The new inference algorithm will generally zero out covariant only ability variables. They don’t serve any purpose.

**Comment by mitchellwrosen** [comment]
This was fixed by #2777
